scipy/stats/_axis_nan_policy.py

```diff
@@ -447,14 +489,22 @@ def hypotest_fun_out(*samples, **kwds):
 
                 if sentinel:
                     samples = _remove_sentinel(samples, paired, sentinel)
-                return hypotest_fun_out(*samples, **kwds)
+                res = hypotest_fun_out(*samples, **kwds)
+                if keepdims:
+                    return _add_reduced_axes(res, reduced_axes, n_outputs,
```

Comment:
It's unfortunate that there are already so many return statements. Consequently, I think you missed (at least) one - original line 435, new line 477.

I tried to limit separate `return`s to fundamentally different cases that would be very awkward to combine, but maybe I could have done better. I'd suggest that philosophy in this PR, though. Can this be restructured so that this `return` turns into a `res = `?

Consider whether you'd prefer to pass `keepdims` as an argument into the function and put the `if` statement into the function rather than always prefacing `_add_reduced_axes` with `if keepdims`.

Reply:
> I think you missed (at least) one - original line 435, new line 477

That return statement just returns a scalar `nan` for each output. As the output there isn't consistent with the shape of other outputs, I wasn't sure how would `keepdims` behave there. I thought it's best to keep it untouched but this needs to be discussed further.

Reply:
> Consider whether you'd prefer to pass `keepdims` as an argument into the function and put the `if` statement into the function rather than always prefacing `_add_reduced_axes` with `if keepdims`

I tried this but `result_unpacker` confuses me. I thought it is supposed to unpack the `result_object` but turns out it is also used to unpack the result returned by `np.apply_along_axis` (see last return). So, I was hesitant to apply `result_unpacker` everywhere before passing it to the `_add_reduced_axes`.

> Can this be restructured so that these new `return`s turn into `res = `?

Done.

Reply:
> That return statement just returns a scalar nan for each output. As the output there isn't consistent with the shape of other outputs, I wasn't sure how would keepdims behave there. 

That part of the decorator starting between `# if axis is not needed, just handle nan_policy and return` and `# check for empty input` is for when all samples are 1D - whether that's because they have been revealed due to the value of `axis` or whether they were passed in as 1D. So yes, currently it returns a scalar `nan` for each output, but that's the correct behavior. It's not inconsistent, because a few lines later, `return hypotest_fun_out(*samples, **kwds)` will return a scalar value for each output because we're still in the part of the decorator where all the samples are 1D. 

Both need to change if we're adding back dimensions.

> this needs to be discussed further.

It definitely does when we apply the decorator to `mode` because currently this assumes that the correct behavior for `propagate` is to return a NaN. That's not true for `mode`. We'll want to add a parameter to the decorator factory like, e.g. `nan_propagate_override=True`. When True (default and current behavior), the decorator doesn't even execute the function if there are NaNs in the slice. When set False, the decorator actually executes the function as usual. I think this will mean that the line:
```python3
if any(contains_nans) and nan_policy == 'propagate':
```
becomes
```python3
if any(contains_nans) and nan_policy == 'propagate' and nan_propagate_override:
```
in the 1D section and similarly in the ND section.

Reply:
> I tried this but result_unpacker confuses me. I thought it is supposed to unpack the result_object but turns out it is also used to unpack the result returned by np.apply_along_axis (see last return).

The idea is that it's written either way. `result_unpacker` ensures that each separate output of the function is an element of a sequence (tuple). It becomes especially important when the object returned by function is not a sequence / cannot be unpacked like a tuple. 

See, for instance, commit f27d887d7107405b613426a5e8301e8476ca6f5b in which I removed the decorator from a bunch of functions I tested it on during development. `CramerVonMisesResults` are not iterable. They are not namedtuples like most stats function returns; they just are objects with fields `statistic` and `value`. The function isn't vectorized, either, so the decorator vectorizes it using `np.vectorize`.  Suppose we call `cramervonmises` on an `M` x `N` input with `axis=0`. `np.vectorize` produces an object array of shape `(N,)` in which each element is a `CramerVonMisesResult`. Before we return it to the user, we need to unpack it and repack it into a single `CramerVonMisesResult` in which `statistic` and `pvalue` are both numerical arrays of shape `(N,)`:
```
result_unpacker=np.vectorize(lambda res: (res.statistic, res.pvalue))
```
This use of `np.vectorize` loops over the object array to pick out the `statistic` and `pvalue`, and the result is a tuple of numerical arrays of shape `(N,)`. Maybe this will help:
```python3
from dataclasses import make_dataclass
Result = make_dataclass('Result', ('statistic', 'pvalue'))
res = np.array([Result(1, 2), Result(3, 4), Result(5, 6)])
res  # array([Result(statistic=1, pvalue=2), Result(statistic=3, pvalue=4), Result(statistic=5, pvalue=6)], dtype=object)
result_unpacker=np.vectorize(lambda res: (res.statistic, res.pvalue))
res2 = result_unpacker(res)
res2  # (array([1, 3, 5]), array([2, 4, 6]))
```
This works to unpack a single `Result` object, too.
```python3
res3 = result_unpacker(Result(1, 2))
res3  # (array(1), array(2))
```
Anyway, hope that helps to understand what it's for. We should write the new code the way we think it should be, without having to tiptoe around what's there.

Reply:
> Anyway, hope that helps to understand what it's for.

Thanks, this was a great explanation! I think I got the purpose of `result_unpacker` and what it's doing.

> We should write the new code the way we think it should be, without having to tiptoe around what's there.

Makes sense. I will rethink this and hopefully simplify all the tiptoeing!

> That part of the decorator starting between `# if axis is not needed, just handle nan_policy and return` and `# check for empty input` is for when all samples are 1D - whether that's because they have been reveled due to the value of `axis` or whether they were passed in as 1D.

Oops, I forgot thers's `if np.all(ndims <= 1)` condition! Will change this and add tests for it.

Reply:
Removed `if keepdims` from everywhere and use `_add_reduced_axes` to remaining `return` statements. Tests for them also added.